cmd/shim/main.go

```diff
@@ -54,6 +61,89 @@ func main() {
 	}
 }
 
+func internalCommand() int {
+	if len(os.Args) < 2 {
+		fmt.Fprintf(os.Stderr, "usage: %s <command> [<args>]\n", os.Args[0])
+		return 1
+	}
+
+	cmd := os.Args[1]
+	args := os.Args[2:]
+
+	switch cmd {
+	case "check":
+		if err := check(args); err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return 1
+		}
+		return 0
+	default:
+		fmt.Fprintf(os.Stderr, "unknown command: %s\n", cmd)
+		return 1
+	}
+}
+
+func check(args []string) error {
+	if len(args) == 0 {
+		return fmt.Errorf("usage: check <host> port/tcp [port/udp ...]")
+	}
+
+	host, ports := args[0], args[1:]
+
+	for _, port := range ports {
+		port, network, ok := strings.Cut(port, "/")
+		if !ok {
+			network = "tcp"
+		}
+
+		pollAddr := net.JoinHostPort(host, port)
+
+		fmt.Println("polling for port", pollAddr)
```

Comment:
nit: if there is an easy way to prepend these prints to have something specific to the container (maybe the hostname?) that will be helpful for future debugging since I'm guessing they currently just all show up in the engine logs together without any context on where they're coming from (unless I missed something somewhere).

No problem doing this in a follow-up or the indefinite future, just seems like it will probably come up eventually.

Reply:
Right now these go nowhere unless you run the engine session with `_DAGGER_DEBUG_HEALTHCHECKS=1`, in which case they'll go to the session output (and directed to the `LogOutput` by the looks of things).

I'd like to get them sent into the event stream along with all the other Buildkit events but haven't found a good way to do that yet. I'm hoping we can clean this up as we figure out how events should work in general.

The logs already include the _target_ container hostname via the `pollAddr` but I'll sneak an `identity.NewID()` in there just to correlate the raw output.

Reply:
Here's how it looks:

```
[check 0sixgl95j0kngw5ot6pxws8qc] polling for port 65NUND9UVER6C.dns.dagger:8000
[check w112dagvaavb9eg260p3pl9hd] polling for port PKI6ANAOO5G9M.dns.dagger:8000
[check 0sixgl95j0kngw5ot6pxws8qc] port not ready: dial tcp: lookup 65NUND9UVER6C.dns.dagger on 10.87.0.1:53: no such host; elapsed: 2.304449ms
[check w112dagvaavb9eg260p3pl9hd] port not ready: dial tcp: lookup PKI6ANAOO5G9M.dns.dagger on 10.87.0.1:53: no such host; elapsed: 2.872764ms
[check x6xzmad3ovdnldghb5ryocygj] polling for port 1BQNFEDIL88RK.dns.dagger:8000
[check x6xzmad3ovdnldghb5ryocygj] port not ready: dial tcp: lookup 1BQNFEDIL88RK.dns.dagger on 10.87.0.1:53: no such host; elapsed: 2.116534ms
[check r8ik41hhaumniat2xvxpuhfnp] polling for port BNQS7HOKINMVE.dns.dagger:8000
[check 4l3v0xrifysnf8nwlkh1bd4s2] polling for port 0MHTHRNJ1QQ9S.dns.dagger:8000
[check r8ik41hhaumniat2xvxpuhfnp] port not ready: dial tcp: lookup BNQS7HOKINMVE.dns.dagger on 10.87.0.1:53: no such host; elapsed: 2.032937ms
[check 4l3v0xrifysnf8nwlkh1bd4s2] port not ready: dial tcp: lookup 0MHTHRNJ1QQ9S.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.539979ms
[check rgubmg15ulrzhk21n0nl39bfk] polling for port J4J57JGPO9I1I.dns.dagger:8000
[check rgubmg15ulrzhk21n0nl39bfk] port not ready: dial tcp: lookup J4J57JGPO9I1I.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.921877ms
[check w112dagvaavb9eg260p3pl9hd] port not ready: dial tcp: lookup PKI6ANAOO5G9M.dns.dagger on 10.87.0.1:53: no such host; elapsed: 85.790264ms
[check 0sixgl95j0kngw5ot6pxws8qc] port not ready: dial tcp: lookup 65NUND9UVER6C.dns.dagger on 10.87.0.1:53: no such host; elapsed: 124.383858ms
#5 resolve docker.io/library/python:latest 0.1s done
#5 CACHED

#4 
[check 9rp0ksdgcw71adcmnstuq4043] polling for port HGC33POUK3VD0.dns.dagger:8000
[check i134eoqtl80mu9k8p0ru8ju8e] polling for port DROBP6LJ2ITRU.dns.dagger:8000
[check 9rp0ksdgcw71adcmnstuq4043] port not ready: dial tcp: lookup HGC33POUK3VD0.dns.dagger on 10.87.0.1:53: no such host; elapsed: 2.175059ms
[check i134eoqtl80mu9k8p0ru8ju8e] port not ready: dial tcp: lookup DROBP6LJ2ITRU.dns.dagger on 10.87.0.1:53: no such host; elapsed: 3.506247ms
[check r8ik41hhaumniat2xvxpuhfnp] port not ready: dial tcp: lookup BNQS7HOKINMVE.dns.dagger on 10.87.0.1:53: no such host; elapsed: 86.150116ms
[check x6xzmad3ovdnldghb5ryocygj] port not ready: dial tcp: lookup 1BQNFEDIL88RK.dns.dagger on 10.87.0.1:53: no such host; elapsed: 128.639472ms
[check 3d0wmk9g495ia9qge1jggief1] polling for port AKHQDD9ISFQDG.dns.dagger:8000
[check 3d0wmk9g495ia9qge1jggief1] port not ready: dial tcp: lookup AKHQDD9ISFQDG.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.821542ms
[check rgubmg15ulrzhk21n0nl39bfk] port not ready: dial tcp: lookup J4J57JGPO9I1I.dns.dagger on 10.87.0.1:53: no such host; elapsed: 88.758681ms
[check 8gjx6namtxe2xrqt991en0xn3] polling for port 3CI7ADE3N20K6.dns.dagger:8000
[check 8gjx6namtxe2xrqt991en0xn3] port not ready: dial tcp 10.87.0.12:8000: connect: connection refused; elapsed: 660.859µs
[check 4l3v0xrifysnf8nwlkh1bd4s2] port not ready: dial tcp: lookup 0MHTHRNJ1QQ9S.dns.dagger on 10.87.0.1:53: no such host; elapsed: 122.379901ms
[check x6xzmad3ovdnldghb5ryocygj] port not ready: dial tcp: lookup 1BQNFEDIL88RK.dns.dagger on 10.87.0.1:53: no such host; elapsed: 218.477814ms
[check w112dagvaavb9eg260p3pl9hd] port not ready: dial tcp: lookup PKI6ANAOO5G9M.dns.dagger on 10.87.0.1:53: no such host; elapsed: 237.695591ms
[check i134eoqtl80mu9k8p0ru8ju8e] port not ready: dial tcp: lookup DROBP6LJ2ITRU.dns.dagger on 10.87.0.1:53: no such host; elapsed: 124.443154ms
[check 0sixgl95j0kngw5ot6pxws8qc] port not ready: dial tcp: lookup 65NUND9UVER6C.dns.dagger on 10.87.0.1:53: no such host; elapsed: 255.667527ms
[check 9rp0ksdgcw71adcmnstuq4043] port not ready: dial tcp: lookup HGC33POUK3VD0.dns.dagger on 10.87.0.1:53: no such host; elapsed: 141.184921ms
[check 3d0wmk9g495ia9qge1jggief1] port not ready: dial tcp: lookup AKHQDD9ISFQDG.dns.dagger on 10.87.0.1:53: no such host; elapsed: 141.658747ms
[check 8gjx6namtxe2xrqt991en0xn3] port is up at 10.87.0.12:8000
[check rgubmg15ulrzhk21n0nl39bfk] port not ready: dial tcp: lookup J4J57JGPO9I1I.dns.dagger on 10.87.0.1:53: no such host; elapsed: 272.741147ms
[check r8ik41hhaumniat2xvxpuhfnp] port not ready: dial tcp: lookup BNQS7HOKINMVE.dns.dagger on 10.87.0.1:53: no such host; elapsed: 289.685057ms
[check x6xzmad3ovdnldghb5ryocygj] port not ready: dial tcp: lookup 1BQNFEDIL88RK.dns.dagger on 10.87.0.1:53: no such host; elapsed: 340.942494ms
[check i134eoqtl80mu9k8p0ru8ju8e] port not ready: dial tcp: lookup DROBP6LJ2ITRU.dns.dagger on 10.87.0.1:53: no such host; elapsed: 228.585851ms
[check 4l3v0xrifysnf8nwlkh1bd4s2] port not ready: dial tcp: lookup 0MHTHRNJ1QQ9S.dns.dagger on 10.87.0.1:53: no such host; elapsed: 342.947108ms
[check 3d0wmk9g495ia9qge1jggief1] port not ready: dial tcp: lookup AKHQDD9ISFQDG.dns.dagger on 10.87.0.1:53: no such host; elapsed: 255.71086ms
#4 0.365 10.87.0.1 - - [22/Feb/2023 20:54:22] "HEAD / HTTP/1.1" 200 -
#4 0.366 10.87.0.1 - - [22/Feb/2023 20:54:22] "GET / HTTP/1.1" 200 -
[check 0sixgl95j0kngw5ot6pxws8qc] port not ready: dial tcp: lookup 65NUND9UVER6C.dns.dagger on 10.87.0.1:53: no such host; elapsed: 483.832343ms
[check 9rp0ksdgcw71adcmnstuq4043] port not ready: dial tcp: lookup HGC33POUK3VD0.dns.dagger on 10.87.0.1:53: no such host; elapsed: 357.033994ms
[check w112dagvaavb9eg260p3pl9hd] port not ready: dial tcp: lookup PKI6ANAOO5G9M.dns.dagger on 10.87.0.1:53: no such host; elapsed: 488.636871ms
[check 4l3v0xrifysnf8nwlkh1bd4s2] port not ready: dial tcp: lookup 0MHTHRNJ1QQ9S.dns.dagger on 10.87.0.1:53: no such host; elapsed: 468.9006ms
[check rgubmg15ulrzhk21n0nl39bfk] port not ready: dial tcp: lookup J4J57JGPO9I1I.dns.dagger on 10.87.0.1:53: no such host; elapsed: 513.169107ms
[check r8ik41hhaumniat2xvxpuhfnp] port not ready: dial tcp: lookup BNQS7HOKINMVE.dns.dagger on 10.87.0.1:53: no such host; elapsed: 545.63307ms
[check i134eoqtl80mu9k8p0ru8ju8e] port not ready: dial tcp: lookup DROBP6LJ2ITRU.dns.dagger on 10.87.0.1:53: no such host; elapsed: 483.742317ms
[check x6xzmad3ovdnldghb5ryocygj] port not ready: dial tcp: lookup 1BQNFEDIL88RK.dns.dagger on 10.87.0.1:53: no such host; elapsed: 608.180931ms
[check 9rp0ksdgcw71adcmnstuq4043] port not ready: dial tcp: lookup HGC33POUK3VD0.dns.dagger on 10.87.0.1:53: no such host; elapsed: 490.95959ms
[check 3d0wmk9g495ia9qge1jggief1] port is up at 10.87.0.18:8000
[check 0sixgl95j0kngw5ot6pxws8qc] port not ready: dial tcp: lookup 65NUND9UVER6C.dns.dagger on 10.87.0.1:53: no such host; elapsed: 744.184106ms
[check rgubmg15ulrzhk21n0nl39bfk] port not ready: dial tcp: lookup J4J57JGPO9I1I.dns.dagger on 10.87.0.1:53: no such host; elapsed: 800.018271ms
[check r8ik41hhaumniat2xvxpuhfnp] port not ready: dial tcp: lookup BNQS7HOKINMVE.dns.dagger on 10.87.0.1:53: no such host; elapsed: 827.663734ms
#4 0.759 10.87.0.1 - - [22/Feb/2023 20:54:22] "HEAD / HTTP/1.1" 200 -
#4 0.761 10.87.0.1 - - [22/Feb/2023 20:54:22] "GET / HTTP/1.1" 200 -
[check w112dagvaavb9eg260p3pl9hd] port not ready: dial tcp: lookup PKI6ANAOO5G9M.dns.dagger on 10.87.0.1:53: no such host; elapsed: 896.246246ms
[check 4l3v0xrifysnf8nwlkh1bd4s2] port not ready: dial tcp: lookup 0MHTHRNJ1QQ9S.dns.dagger on 10.87.0.1:53: no such host; elapsed: 879.260002ms
[check i134eoqtl80mu9k8p0ru8ju8e] port not ready: dial tcp: lookup DROBP6LJ2ITRU.dns.dagger on 10.87.0.1:53: no such host; elapsed: 823.500875ms
[check 9rp0ksdgcw71adcmnstuq4043] port not ready: dial tcp 10.87.0.19:8000: connect: connection refused; elapsed: 876.964232ms
[check x6xzmad3ovdnldghb5ryocygj] port not ready: dial tcp: lookup 1BQNFEDIL88RK.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.096979631s
[check rgubmg15ulrzhk21n0nl39bfk] port not ready: dial tcp: lookup J4J57JGPO9I1I.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.09932358s
[check w112dagvaavb9eg260p3pl9hd] port not ready: dial tcp: lookup PKI6ANAOO5G9M.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.200586361s
[check r8ik41hhaumniat2xvxpuhfnp] port not ready: dial tcp: lookup BNQS7HOKINMVE.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.161522837s
[check 0sixgl95j0kngw5ot6pxws8qc] port not ready: dial tcp: lookup 65NUND9UVER6C.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.368759271s
[check 9rp0ksdgcw71adcmnstuq4043] port is up at 10.87.0.19:8000
[check 4l3v0xrifysnf8nwlkh1bd4s2] port not ready: dial tcp: lookup 0MHTHRNJ1QQ9S.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.432899891s
[check x6xzmad3ovdnldghb5ryocygj] port not ready: dial tcp: lookup 1BQNFEDIL88RK.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.500625668s
[check i134eoqtl80mu9k8p0ru8ju8e] port not ready: dial tcp: lookup DROBP6LJ2ITRU.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.378859612s
#4 1.465 10.87.0.1 - - [22/Feb/2023 20:54:23] "HEAD / HTTP/1.1" 200 -
#4 1.466 10.87.0.1 - - [22/Feb/2023 20:54:23] "GET / HTTP/1.1" 200 -
[check rgubmg15ulrzhk21n0nl39bfk] port not ready: dial tcp: lookup J4J57JGPO9I1I.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.648003188s
[check w112dagvaavb9eg260p3pl9hd] port not ready: dial tcp: lookup PKI6ANAOO5G9M.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.73601572s
[check 0sixgl95j0kngw5ot6pxws8qc] port not ready: dial tcp: lookup 65NUND9UVER6C.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.888710601s
[check r8ik41hhaumniat2xvxpuhfnp] port not ready: dial tcp: lookup BNQS7HOKINMVE.dns.dagger on 10.87.0.1:53: no such host; elapsed: 1.868415463s
[check i134eoqtl80mu9k8p0ru8ju8e] port is up at 10.87.0.20:8000
#4 2.001 10.87.0.1 - - [22/Feb/2023 20:54:23] "HEAD / HTTP/1.1" 200 -
#4 2.002 10.87.0.1 - - [22/Feb/2023 20:54:23] "GET / HTTP/1.1" 200 -
[check rgubmg15ulrzhk21n0nl39bfk] port is up at 10.87.0.21:8000
[check 4l3v0xrifysnf8nwlkh1bd4s2] port not ready: dial tcp: lookup 0MHTHRNJ1QQ9S.dns.dagger on 10.87.0.1:53: no such host; elapsed: 2.346197911s
#4 2.357 10.87.0.1 - - [22/Feb/2023 20:54:24] "HEAD / HTTP/1.1" 200 -
#4 2.358 10.87.0.1 - - [22/Feb/2023 20:54:24] "GET / HTTP/1.1" 200 -
[check x6xzmad3ovdnldghb5ryocygj] port not ready: dial tcp: lookup 1BQNFEDIL88RK.dns.dagger on 10.87.0.1:53: no such host; elapsed: 2.544055156s
[check 0sixgl95j0kngw5ot6pxws8qc] port not ready: dial tcp: lookup 65NUND9UVER6C.dns.dagger on 10.87.0.1:53: no such host; elapsed: 2.753063649s
[check r8ik41hhaumniat2xvxpuhfnp] port is up at 10.87.0.22:8000
#4 2.939 10.87.0.1 - - [22/Feb/2023 20:54:24] "HEAD / HTTP/1.1" 200 -
#4 2.940 10.87.0.1 - - [22/Feb/2023 20:54:24] "GET / HTTP/1.1" 200 -
[check w112dagvaavb9eg260p3pl9hd] port not ready: dial tcp: lookup PKI6ANAOO5G9M.dns.dagger on 10.87.0.1:53: no such host; elapsed: 3.314189127s
[check 0sixgl95j0kngw5ot6pxws8qc] port not ready: dial tcp: lookup 65NUND9UVER6C.dns.dagger on 10.87.0.1:53: no such host; elapsed: 3.637948546s
[check 4l3v0xrifysnf8nwlkh1bd4s2] port is up at 10.87.0.23:8000
#4 3.824 10.87.0.1 - - [22/Feb/2023 20:54:25] "HEAD / HTTP/1.1" 200 -
#4 3.825 10.87.0.1 - - [22/Feb/2023 20:54:25] "GET / HTTP/1.1" 200 -
[check x6xzmad3ovdnldghb5ryocygj] port not ready: dial tcp: lookup 1BQNFEDIL88RK.dns.dagger on 10.87.0.1:53: no such host; elapsed: 4.017408221s
[check w112dagvaavb9eg260p3pl9hd] port not ready: dial tcp: lookup PKI6ANAOO5G9M.dns.dagger on 10.87.0.1:53: no such host; elapsed: 4.785775442s
[check x6xzmad3ovdnldghb5ryocygj] port not ready: dial tcp: lookup 1BQNFEDIL88RK.dns.dagger on 10.87.0.1:53: no such host; elapsed: 6.710970788s
[check w112dagvaavb9eg260p3pl9hd] port not ready: dial tcp: lookup PKI6ANAOO5G9M.dns.dagger on 10.87.0.1:53: no such host; elapsed: 7.399565902s
[check 0sixgl95j0kngw5ot6pxws8qc] port is up at 10.87.0.24:8000
#4 7.478 10.87.0.1 - - [22/Feb/2023 20:54:29] "HEAD / HTTP/1.1" 200 -
#4 7.479 10.87.0.1 - - [22/Feb/2023 20:54:29] "GET / HTTP/1.1" 200 -
[check x6xzmad3ovdnldghb5ryocygj] port is up at 10.87.0.25:8000
#4 9.032 10.87.0.1 - - [22/Feb/2023 20:54:31] "HEAD / HTTP/1.1" 200 -
#4 9.034 10.87.0.1 - - [22/Feb/2023 20:54:31] "GET / HTTP/1.1" 200 -
[check w112dagvaavb9eg260p3pl9hd] port is up at 10.87.0.26:8000
#4 ...
```